app/src/main/java/nl/eduvpn/app/service/ConnectionService.java

```diff
@@ -188,39 +192,70 @@ public void onTokenRequestCompleted(TokenResponse tokenResponse, AuthorizationEx
     /**
      * Process the the exchanged token.
      *
-     * @param tokenResponse The response from the token exchange.
-     * @param activity      The current activity.
+     * @param authorizationResponse The authorization response.
+     * @param tokenResponse         The response from the token exchange updated into the authentication state
+     * @param activity              The current activity.
      */
-    private void _processTokenExchangeResponse(TokenResponse tokenResponse, Activity activity) {
-        String accessToken = tokenResponse.accessToken;
+    private void _processTokenExchangeResponse(AuthorizationResponse authorizationResponse, TokenResponse tokenResponse, Activity activity) {
+        AuthState authState = new AuthState(authorizationResponse, tokenResponse, null);
+        String accessToken = authState.getAccessToken();
         if (accessToken == null || accessToken.isEmpty()) {
             ErrorDialog.show(activity, R.string.error_dialog_title, R.string.error_access_token_missing);
             return;
         }
-        // Save the access token
-        _accessToken = accessToken;
-        _preferencesService.storeCurrentAccessToken(accessToken);
+        // Save the authentication state.
+        _preferencesService.storeCurrentAuthState(authState);
         // Save the access token for later use.
-        _historyService.cacheAccessToken(_preferencesService.getCurrentInstance(), _accessToken);
+        _historyService.cacheAuthenticationState(_preferencesService.getCurrentInstance(), authState);
         Toast.makeText(activity, R.string.provider_added_new_configs_available, Toast.LENGTH_LONG).show();
     }
 
     /**
-     * Returns the access token which authenticates against the current API.
+     * Returns a single which emits a fresh access token which is to be used with the API.
      *
-     * @return The access token used to authenticate.
+     * @return The access token used to authenticate in an emitter.
      */
-    public String getAccessToken() {
-        return _accessToken;
+    public PublishSubject<String> getFreshAccessToken() {
+        final PublishSubject<String> publishSubject = PublishSubject.create();
```

Comment:
Why create an Subject everytime? Shouldn't this Subject be in an field so you have only one instance and can have multiple subscribers?
If you just need an single result use `Single.fromCallable()` or `Single.just()`. Single will terminate after 1 item.

Reply:
You are right, converted to Singles.